### ISSUE TITLE
Use autoLoad to speed up loading code

### ISSUE THREAD

**Issue by bdewater** [plaid/plaid-ruby#454]
The Plaid gem is quite heavy with over 1000 model files in `lib/plaid/models`. These are all forcibly loaded here: https://github.com/plaid/plaid-ruby/blob/9143e242edb4dd5f77e6d859bed5ca24d4aece60/lib/plaid.rb#L20-L1091 

If this were to use Ruby's autoload feature it would only load files on demand, speeding up the time it takes to start a Rake task or Rails app (and lowering memory usage). Fortunately the OpenAPI generator supports this, see:
- https://github.com/OpenAPITools/openapi-generator/issues/12648
- https://github.com/OpenAPITools/openapi-generator/pull/13153

Could this be implemented?

**Comment by phoenixy1** [comment]
@bdewater thank you for the suggestion! It looks like this functionality was added in OpenAPI generator 6.1 and we are currently on 5.1.1. I know the team has been looking into migrating our libraries to use version 6, but it's turned out to be a non-trivial project and not on the roadmap for H1. I've logged a task in our internal tracker so we can track this as both yet another reason to upgrade and to have it on the task list for when we do upgrade.

**Comment by maxwell** [comment]
To +1 this; the `plaid` gem is currently the number 1 largest gem in my mature 10 y/o application, surpassing even rails.  It is almost 30% of the total memory on boot of my application.

`
TOP: 178.3906 MiB
  plaid: 55.8594 MiB
    plaid/api/plaid_api: 3.8125 MiB
  rails/all: 53.1094 MiB
```

**Comment by phoenixy1** [comment]
Just to follow up on this, we've upgraded to auto-generator 6.1, for this release, which is a pre-requisite for enabling autoload. I'm going to try enabling it and if all goes well, autoload will be enabled in our next ruby client library release

**Comment by phoenixy1** [comment]
Following up again, I'm currently at the stage of vetting this proposal with our internal ruby expert, and he asked if you could let us know what kind of load time you're seeing on the Plaid gem? Getting this info will help us understand the impact of not using autoload and the risk/reward payoff of enabling it.

**Comment by phoenixy1** [comment]
@maxwell @bdewater I just wanted to follow up on the questions above? If you could provide this information, that would really help us get this change reviewed and approved. Thanks!

**Comment by maxwell** [comment]
@phoenixy1 thank you for following up.

That's what I have right now given my time constraints.  What I can say is that my app uses over 170 gems, (including Rails) and Plaid represents 30% of memory of my app on boot, so it _feels_ like maybe the gem is not doing some sort of best practice, as I also use a fair amount of involved clients of 3rd party services. Given I only use the Gem in a few places, I guess my alternative is to just NOT use it and roll my own, but it _feels_ like maybe a smell that autoloading would resolve.  I'm not an expert in Gem packaging, but I do know many gems follow a pattern where including the gem does not immediately pull in EVERY file into memory, and follow more of the lazy autoload pattern.

**Comment by bdewater** [comment]
@phoenixy1 I'm currently on leave, but IIRC after adding the Plaid gem I saw ~1s increase in app cold boot in our Rails app in development environment.

**Comment by phoenixy1** [comment]
BTW, this didn't make today's release, but I did get approval and plan to get it out for next month.

**Comment by seanbjornsson** [comment]
Hey @phoenixy1 , did this make it into a release?

**Comment by phoenixy1** [comment]
@seanbjornsson sorry for the delay and thank you for the reminder! I got super busy right around then with some other projects, but I'm making another stab at adding this and currently have a PR awaiting review that I think should fix it.

**Comment by bdewater** [comment]
I forgot to circle back. Quick benchmark:
```rb
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "memory_profiler"
  gem "rails", "7.1.0", require: false
  gem "plaid", "23.0.0", require: false
end

report = MemoryProfiler.report do
  require "rails/all"
  require "plaid"
end

report.pretty_print
```

Run with `ruby benchmark.rb`. Some results that stand out:

Memory and objects allocated are an order of magnitude more than anything else:
```
allocated memory by gem
-----------------------------------
  62921889  plaid-23.0.0
   5845889  mail-2.8.1
   4659426  activesupport-7.1.0
   3172977  concurrent-ruby-1.2.2
   2349455  activerecord-7.1.0
   2324454  nokogiri-1.15.4-arm64-darwin
   1453437  tzinfo-2.0.6
   1409491  actionpack-7.1.0
   1312433  marcel-1.0.2
   1064948  rubygems
[omitted rest of list]

allocated objects by gem
-----------------------------------
    692895  plaid-23.0.0
     53900  mail-2.8.1
     42565  activesupport-7.1.0
     28782  concurrent-ruby-1.2.2
     22197  nokogiri-1.15.4-arm64-darwin
     21635  activerecord-7.1.0
     19408  marcel-1.0.2
     13801  actionpack-7.1.0
     12044  tzinfo-2.0.6
     10062  rack-3.0.8
[omitted rest of list]
```
Same for retained:
```
retained memory by gem
-----------------------------------
   6112398  plaid-23.0.0
    484035  activesupport-7.1.0
    483754  marcel-1.0.2
    432669  rubygems
    310861  nokogiri-1.15.4-arm64-darwin
    299503  activerecord-7.1.0
    255463  concurrent-ruby-1.2.2
    239713  mail-2.8.1
    233470  actionpack-7.1.0
    205381  rack-3.0.8
    139993  railties-7.1.0
    108081  tzinfo-2.0.6
[omitted rest of list]

retained objects by gem
-----------------------------------
     31283  plaid-23.0.0
      8387  marcel-1.0.2
      4557  activesupport-7.1.0
      2737  activerecord-7.1.0
      2594  nokogiri-1.15.4-arm64-darwin
      2210  concurrent-ruby-1.2.2
      2082  rack-3.0.8
      2035  rubygems
      1840  actionpack-7.1.0
      1790  mail-2.8.1
      1347  railties-7.1.0
[omitted rest of list]
```

Removing the memory_profiler reporting and calling `time ruby benchmark.rb`:
- with `require "plaid"`: 1.87s average on my machine (Apple M1 Pro 14")
- without `require "plaid"`: 1.31s average

Note that Rails 7.1 and it's dependencies comprise 65 gems, so half a second for Plaid alone is a lot.

**Comment by phoenixy1** [comment]
Just following up, the PR that builds the library with useAutoload=true has just been merged, so this should be incorporated into the next client library release

**Comment by javierjulio** [comment]
@bdewater thanks for sharing that script. I ran that first with Plaid 22.0.0, then with Plaid 24.2.0 and noticed a significant decrease in allocated and retained memory for the gem.

**Comment by phoenixy1** [comment]
This has now been shipped in 24.0.0, so I'm going to close this out. Thanks all who helped report and profile this issue!
